### PR TITLE
feat(angular): support binding routing data to component inputs

### LINKED ISSUE #27476: feat: angular, support binding route params to component inputs
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already report this problem, without success.

### Ionic Framework Version

v6.x, v7.x

### Current Behavior

In Angular v16, they added a new feature to help Enables binding information from the [Router](https://angular.io/api/router/Router) state directly to the inputs of the component in [Route](https://angular.io/api/router/Route) configurations.

When I add the following code as mentioned [here](https://angular.io/api/router/withComponentInputBinding) 

I got this alert in console : 

**'withComponentInputBinding' feature is enabled but this application is using an outlet that may not support binding to component inputs.**

Screenshot 2023-05-15 at 3 57 42 PM

### Expected Behavior

I am excepting to get router info  from the input tag `@Input`

### Steps to Reproduce

``` javascript

import { provideRouter, Routes, withComponentInputBinding } from '@angular/router';
const routes: Routes = [....]
export const ROUTE_PROVIDERS = provideRouter(routes, withComponentInputBinding());

or 


@NgModule({
  imports: [
    RouterModule.forRoot(routes, {
      //... other features
      bindToComponentInputs: true // <-- enable this feature
    })
  ],
})
export class AppModule {}

```

### Code Reproduction URL

[Editor Stackbiltz](https://stackblitz.com/edit/ionic6-angular13-x8k9d4)
[Preview Stackbiltz](https://ionic6-angular13-x8k9d4.stackblitz.io)

### Ionic Info

```
Ionic:
   Ionic Framework               : @ionic/angular 7.0.6
   @angular-devkit/build-angular : 16.0.1
   @angular-devkit/schematics    : 16.0.1
   @angular/cli                  : 16.0.1
   @ionic/angular-toolkit        : 9.0.0


```

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by sean-perkins**
Issue number: Resolves #27476

---------

## What is the current behavior?

Ionic Angular application on Angular v16 cannot use the [`bindToComponentInputs`](https://angular.io/api/router/ExtraOptions#bindToComponentInputs) feature to assign route parameters, query parameters, route data and route resolve data to component inputs.

## What is the new behavior?

- Ionic Angular developers can use the option on `RouterModule.forRoot` to enable the Angular feature for binding the route snapshot data to the component inputs.

**Modules**
```ts
@NgModule({
  imports: [
    RouterModule.forRoot([/* your routes */], { 
      bindToComponentInputs: true // <-- enable this feature
    })
  ]
})
export class AppModule { }
```

**Standalone**

```ts
import { withComponentInputBinding } from '@angular/router';

bootstrapApplication(App, {
  providers: [
    provideRouter(routes, 
      //... other features
      withComponentInputBinding() // <-- enable this feature
    )
  ],
});
```

With this feature enabled, developers can bind route parameters, query parameters, route data and the returned value from a resolver to input bindings on their component. 

For example, with a route configuration of:
```ts
RouterModule.forChild([
  {
    path: ':id',
    data: {
      title: 'Hello world'
    },
    resolve: {
      name: () => 'Resolved name'
    },
    loadComponent: () => import('./example-component/example.component').then(c => c.ExampleComponent)
  }
])
```
and a component configuration of:

```ts
@Component({ }) 
export class ExampleComponent {
  @Input() id?: string;
  @Input() title?: string;
  @Input() name?: string;
  @Input() query?: string;
}
```

Navigating to the component with a url of: `/2?query=searchphrase`

The following would occur:
- `id` would return `2`
- `title` would return `Hello world`
- `name` would return `Resolved name`
- `query` would return `searchphrase`

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

This PR will need to be targeted to a minor release once a design doc is approved by the team.

Dev-build: `7.1.3-dev.11689276547.129acb40`

### CODE REVIEW

File: packages/angular/src/directives/navigation/ion-router-outlet.ts [hunk 1]
```diff
@@ -39,22 +43,30 @@ import { RouteView, getUrl } from './stack-utils';
...
+  /** @internal */
+  get activatedComponentRef(): ComponentRef<any> | null {
+    return this.activated;
+  }
+  private _activatedRoute: ActivatedRoute | null = null;
+
+  /**
+   * The name of the outlet
+   *
+   * @see [named outlets](guide/router-tutorial-toh#displaying-multiple-routes-in-named-outlets)
```
Comment (liamdebeasi) [inline packages/angular/src/directives/navigation/ion-router-outlet.ts:67]: Where is this supposed to link to?

File: packages/angular/src/directives/navigation/ion-router-outlet.ts [hunk 2]
```diff
@@ -415,3 +433,93 @@ class OutletInjector implements Injector {
...
+      });
+
+    this.outletDataSubscriptions.set(outlet, dataSubscription);
+  }
+}
+
+/**
+ * The list of features as an enum to uniquely type each feature.
+ */
+export const enum RouterFeatureKind {
```
Comment (liamdebeasi) [inline packages/angular/src/directives/navigation/ion-router-outlet.ts:515]: I don't see this being used anywhere.

File: packages/angular/src/directives/navigation/ion-router-outlet.ts [hunk 3]
```diff
@@ -415,3 +433,93 @@ class OutletInjector implements Injector {
     return this.parent.get(token, notFoundValue);
   }
 }
+
+export const INPUT_BINDER = new InjectionToken<RoutedComponentInputBinder>('');
```
Comment (liamdebeasi) [inline packages/angular/src/directives/navigation/ion-router-outlet.ts:437]: Can we add TODOs/a spike in our backlog so we know to remove this once Angular <16 is eventually dropped?

File: packages/angular/test/apps/ng16/src/app/version-test/bind-component-inputs/bind-component-inputs.component.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    <ion-header>
+      <ion-toolbar>
+        <ion-title>bindToComponentInputs</ion-title>
+    </ion-toolbar>
+    </ion-header>
+    <ion-content class="ion-padding">
+      <div>
+        <h3>Bind Route</h3>
+        <p id="route-params">Route params: id: {{id}}</p>
+        <p id="query-params">Query params: query: {{query}}</p>
```
Comment (liamdebeasi) [inline packages/angular/test/apps/ng16/src/app/version-test/bind-component-inputs/bind-component-inputs.component.ts:16]: The query params don't get set when I do `http://localhost:4200/version-test/bind-route/123?liam=hello`

Comment (sean-perkins) [reply packages/angular/test/apps/ng16/src/app/version-test/bind-component-inputs/bind-component-inputs.component.ts:16]: For that example to work, you would need to add:
```ts
@Input() liam?: string;
```

to the component class & render it in the template. With the current example, the query parameter would be mapped with a URL like: `http://localhost:4200/version-test/bind-route/123?query=hello`

File: packages/angular/src/directives/navigation/ion-router-outlet.ts (hunk 1 above)
Comment (sean-perkins) [reply packages/angular/src/directives/navigation/ion-router-outlet.ts:67]: This is copy/paste from Angular's source. I'll remove that extra link, but keep the comment block.

**Review by liamdebeasi — APPROVED** [review]
Tested this out, and it works well. Great job!

**Pull request merged**